qiskit_experiments/library/tomography/basis/local_basis.py

```diff
@@ -0,0 +1,413 @@
+                            defined in the qubit_states.
+            qubit_states: Optional, density matrices prepared by the input
+                          instructions on specified qubits. If None the default
+                          states will be used for all qubits.
+
+        Raises:
+            QiskitError: if input states or instructions are not valid.
+        """
+        if instructions is None and default_states is None and qubit_states is None:
+            raise QiskitError(
+                "LocalPreparationBasis must define at least one of instructions, "
+                "default_states, or qubit_states."
+            )
+        super().__init__(name)
+
+        # Internal variables
+        self._instructions = tuple()
+        self._size = None
+        self._default_states = None
+        self._default_dim = None
+        self._qubit_states = {}
+        self._qubit_dim = {}
+        self._qubits = set()
+        self._custom_defaults = True
+
+        # Format instructions so compatible types can be converted to
+        # Instruction instances.
+        if instructions is not None:
+            self._instructions = _format_instructions(instructions)
+            self._size = len(instructions)
+            if default_states is None:
+                default_states = self._instructions
+                self._custom_defaults = False
+
+        # Construct default states
+        if default_states is not None:
+            self._default_states = tuple(DensityMatrix(i).data for i in default_states)
+            self._default_dim = self._default_states[0].shape[0]
+            if self._size is None:
+                self._size = len(self._default_states)
+            elif len(self._default_states) != self._size:
+                raise QiskitError(
+                    "Number of instructions and number of default states must be equal."
+                )
+
+        # Construct states of specific qubits if provided
+        qubit_states = qubit_states or {}
+        for qubit, states in qubit_states.items():
+            if self._size is None:
+                self._size = len(states)
+            elif len(states) != self._size:
+                raise QiskitError("Number of instructions and number of states must be equal.")
+
+            qstates = tuple(DensityMatrix(i).data for i in states)
+            self._qubit_states[qubit] = qstates
+            self._qubit_dim[qubit] = qstates[0].shape[0]
+            self._qubits.add(qubit)
+
+        # Pseudo hash value to make basis hashable for LRU cached functions
+        self._hash = hash(
+            (
+                type(self),
+                self._name,
+                self._size,
+                self._default_dim,
+                self._custom_defaults,
+                tuple(self._qubits),
+                tuple(self._qubit_dim.values()),
+                (type(i) for i in self._instructions),
```

Comment:
This is just to implement a basic pseudo-hash function so these classes can work with lru_caches like was done with the old bases classes. You can't hash an actual instruction instance because they are too complicated.

Reply:
Fair enough.